File: frontend/lib/adaptive/bkt.ts
/**
 * Bayesian Knowledge Tracing (BKT) Service
 * =======================================
 * Estimates student mastery probability for knowledge concepts.
 */

export class BKTParameters {
  public priorLearned: number;
  public transitionLearn: number;
  public guess: number;
  public slip: number;

  constructor(
    priorLearned: number = 0.25,
    transitionLearn: number = 0.06,
    guess: number = 0.20,
    slip: number = 0.10
  ) {
    this.priorLearned = priorLearned;
    this.transitionLearn = transitionLearn;
    this.guess = guess;
    this.slip = slip;
  }
}

/**
 * Tính xác suất hậu nghiệm P(L_t | Result) sau khi trả lời Đúng/Sai.
 */
export function calculateBktPosterior(
  pMastery: number,
  isCorrect: boolean,
  params: BKTParameters
): number {
  let numerator = 0;
  let denominator = 0;

  if (isCorrect) {
    numerator = pMastery * (1.0 - params.slip);
    denominator = numerator + (1.0 - pMastery) * params.guess;
  } else {
    numerator = pMastery * params.slip;
    denominator = numerator + (1.0 - pMastery) * (1.0 - params.guess);
  }

  if (denominator === 0) {
    return pMastery;
  }

  return Math.min(1.0, Math.max(0.0, numerator / denominator));
}

/**
 * Cập nhật xác suất làm chủ của học sinh sau khi làm câu hỏi với điểm actualScore.
 * Hỗ trợ partial credit bằng nội suy tuyến tính.
 * Giới hạn [0.0001, 0.9999] để tránh Mastery Trap.
 */
export function calculateBktUpdate(
  pMastery: number,
  actualScore: number,
  params: BKTParameters
): number {
  let pPosterior = 0;

  if (actualScore >= 1.0) {
    pPosterior = calculateBktPosterior(pMastery, true, params);
  } else if (actualScore <= 0.0) {
    pPosterior = calculateBktPosterior(pMastery, false, params);
  } else {
    // Nội suy tuyến tính
    const pPostCorrect = calculateBktPosterior(pMastery, true, params);
    const pPostIncorrect = calculateBktPosterior(pMastery, false, params);
    pPosterior = actualScore * pPostCorrect + (1.0 - actualScore) * pPostIncorrect;
  }

  // Áp dụng học tập chuyển tiếp P(T)
  const pMasteryNew = pPosterior + (1.0 - pPosterior) * params.transitionLearn;

  // Giới hạn [0.0001, 0.9999] và làm tròn 4 chữ số thập phân
  const clamped = Math.min(0.9999, Math.max(0.0001, pMasteryNew));
  return Math.round(clamped * 10000) / 10000;
}

/**
 * Ánh xạ xác suất BKT sang nhãn trạng thái tương ứng trong Database:
 * - Mastery 1 (Weak): P(L) < 0.30
 * - Mastery 2 & 3 (Learning): 0.30 <= P(L) < 0.85
 * - Mastery 4 (Mastered): P(L) >= 0.85
 */
export function determineMasteryState(pMastery: number): string {
  if (pMastery < 0.30) {
    return 'weak';
  } else if (pMastery < 0.85) {
    return 'learning';
  } else {
    return 'mastered';
  }
}

File: src/services/adaptive/bkt.py
"""
Bayesian Knowledge Tracing (BKT) Service
=======================================

Module này thực hiện mô hình toán học Bayesian Knowledge Tracing để ước lượng
xác suất làm chủ tri thức (mastery probability) của học sinh đối với từng Concept.

Tài liệu tham chiếu thiết kế:
---------------------------
- docs/research/bayesian-knowledge-tracing.md
- docs/research/adaptive-implementation_guide.md
"""


class BKTParameters:
    def __init__(
        self,
        prior_learned: float = 0.25,
        transition_learn: float = 0.06,
        guess: float = 0.20,
        slip: float = 0.10,
    ):
        """
        Khởi tạo tham số BKT cho một concept kiến thức.
        - prior_learned (P(L0)): Xác suất ban đầu học sinh đã làm chủ kiến thức.
        - transition_learn (P(T)): Xác suất học sinh học được sau mỗi cơ hội học tập.
        - guess (P(G)): Xác suất đoán bừa đúng dù chưa làm chủ kiến thức.
        - slip (P(S)): Xác suất làm sai ngớ ngẩn dù đã làm chủ kiến thức.
        """
        self.prior_learned = prior_learned
        self.transition_learn = transition_learn
        self.guess = guess
        self.slip = slip


def calculate_bkt_posterior(
    p_mastery: float,
    is_correct: bool,
    params: BKTParameters,
) -> float:
    """
    Tính xác suất hậu nghiệm (Posterior Probability) P(L_t | Result)
    ngay sau khi nhận kết quả câu trả lời Đúng/Sai.
    """
    if is_correct:
        numerator = p_mastery * (1.0 - params.slip)
        denominator = numerator + (1.0 - p_mastery) * params.guess
    else:
        numerator = p_mastery * params.slip
        denominator = numerator + (1.0 - p_mastery) * (1.0 - params.guess)

    if denominator == 0:
        return p_mastery
    return min(1.0, max(0.0, numerator / denominator))


def calculate_bkt_update(
    p_mastery: float,
    actual_score: float,
    params: BKTParameters,
) -> float:
    """
    Cập nhật và trả về xác suất làm chủ mới của học sinh sau lượt thử nghiệm.
    Sử dụng logic nhị phân chuẩn (Đúng khi actual_score >= 0.75, ngược lại là Sai).
    """
    is_correct = actual_score >= 0.75
    p_posterior = calculate_bkt_posterior(p_mastery, is_correct, params)

    # 2. Áp dụng xác suất học tập chuyển tiếp P(T) cho cơ hội học tiếp theo
    p_mastery_new = p_posterior + (1.0 - p_posterior) * params.transition_learn

    # Giới hạn cận trên là 0.9999 và cận dưới là 0.0001, làm tròn 4 chữ số
    return round(min(0.9999, max(0.0001, p_mastery_new)), 4)


def determine_mastery_state(p_mastery: float) -> str:
    """
    Ánh xạ xác suất BKT của học viên về trạng thái tương ứng trong Database:
    - Mastery 1 (Weak): P(L) < 0.30
    - Mastery 2 & 3 (Learning): 0.30 <= P(L) < 0.85
    - Mastery 4 (Mastered): P(L) >= 0.85
    """
    if p_mastery < 0.30:
        return "weak"
    elif p_mastery < 0.85:
        return "learning"
    else:
        return "mastered"
